src/sim/mem_pool.cc

```diff
@@ -114,31 +95,70 @@ MemPool::totalBytes() const
 Addr
 MemPool::allocate(Addr npages)
 {
-    Addr return_addr = freePageAddr();
-    freePageNum += npages;
+    Addr page;
+    if (!freePhysPages.allocate(npages, page))
+        fatal("Out of memory, please increase size of physical memory.");
 
-    fatal_if(freePages() <= 0,
-            "Out of memory, please increase size of physical memory.");
+    return page << pageShift;
+}
 
-    return return_addr;
+void
+MemPool::deallocate(Addr dealloc_start, Addr dealloc_npages)
+{
+    freePhysPages.insert(dealloc_start >> pageShift, dealloc_npages);
 }
 
 void
 MemPool::serialize(CheckpointOut &cp) const
 {
     paramOut(cp, "page_shift", pageShift);
     paramOut(cp, "start_page", startPageNum);
-    paramOut(cp, "free_page_num", freePageNum);
     paramOut(cp, "total_pages", _totalPages);
+
+    ScopedCheckpointSection sec(cp, "free_list");
+    paramOut(cp, "size", freePhysPages.ranges().size());
+    int i = 0;
+    for (const auto& range : freePhysPages.ranges()) {
+        ScopedCheckpointSection sec(cp, csprintf("free%d", i++));
+        paramOut(cp, "base", range.base);
+        paramOut(cp, "size", range.size);
+    }
 }
 
 void
 MemPool::unserialize(CheckpointIn &cp)
 {
     paramIn(cp, "page_shift", pageShift);
     paramIn(cp, "start_page", startPageNum);
-    paramIn(cp, "free_page_num", freePageNum);
     paramIn(cp, "total_pages", _totalPages);
+
+    Counter free_page_num;
+    if (optParamIn(cp, "free_page_num", free_page_num)) {
+        // Handle checkpoints generated by legacy page allocator
+        // for backwards compatibility.
+        warn("Checkpoint contains legacy MemPool state (free_page_num)\n");
+        freePhysPages.insert(free_page_num, _totalPages - free_page_num);
+    } else {
+        ScopedCheckpointSection sec(cp, "free_list");
+        int n;
```

Comment:
```suggestion
        int n = 0;
```

Just to make sure the static analysis tools won't report this as an uninitialized variable.

Reply:
Fixed!